src/spotify_to_tidal/sync.py

```diff
@@ -90,27 +97,34 @@ def match(tidal_track, spotify_track) -> bool:
         and artist_match(tidal_track, spotify_track)
     )
 
-def tidal_search(spotify_track_and_cache, tidal_session: tidalapi.Session) -> tidalapi.Track | None:
-    spotify_track, cached_tidal_track = spotify_track_and_cache
-    if cached_tidal_track: return cached_tidal_track
-    if spotify_track['id'] is None: return None
-    if failure_cache.has_match_failure(spotify_track['id']):
-        return None
-    # search for album name and first album artist
-    if 'album' in spotify_track and 'artists' in spotify_track['album'] and len(spotify_track['album']['artists']):
-        album_result = tidal_session.search(simple(spotify_track['album']['name']) + " " + simple(spotify_track['album']['artists'][0]['name']), models=[tidalapi.album.Album])
-        for album in album_result['albums']:
-            album_tracks = album.tracks()
-            if len(album_tracks) >= spotify_track['track_number']:
-                track = album_tracks[spotify_track['track_number'] - 1]
-                if match(track, spotify_track):
-                    failure_cache.remove_match_failure(spotify_track['id'])
-                    return track
-    # if that fails then search for track name and first artist
-    for track in tidal_session.search(simple(spotify_track['name']) + ' ' + simple(spotify_track['artists'][0]['name']), models=[tidalapi.media.Track])['tracks']:
-        if match(track, spotify_track):
-            failure_cache.remove_match_failure(spotify_track['id'])
-            return track
+async def tidal_search(spotify_track, rate_limiter, tidal_session: tidalapi.Session) -> tidalapi.Track | None:
+    def _search_for_track_in_album():
+        # search for album name and first album artist
+        if 'album' in spotify_track and 'artists' in spotify_track['album'] and len(spotify_track['album']['artists']):
+            album_result = tidal_session.search(simple(spotify_track['album']['name']) + " " + simple(spotify_track['album']['artists'][0]['name']), models=[tidalapi.album.Album])
+            for album in album_result['albums']:
+                album_tracks = album.tracks()
+                if len(album_tracks) >= spotify_track['track_number']:
+                    track = album_tracks[spotify_track['track_number'] - 1]
+                    if match(track, spotify_track):
+                        failure_cache.remove_match_failure(spotify_track['id'])
+                        return track
+    def _search_for_standalone_track():
```

Comment:
No changes to the implementation here, the code has just been moved into a nested function